File: 09_02_2024_Largest_Divisible_Subset.py
from typing import List

class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        n = len(nums)
        nums.sort()
        dp = [[] for _ in range(n)]
        dp[n-1] = [nums[n-1]]
        ans_size = 1
        ans_index = n-1
        for i in range(n-2, -1, -1):
            sub_ans_size = 0
            sub_ans_index = -1
            for j in range(i+1, n):
                if nums[j] % nums[i] == 0 and len(dp[j]) > sub_ans_size:
                    sub_ans_size = len(dp[j])
                    sub_ans_index = j
            dp[i] = [nums[i]]
            if sub_ans_index != -1:
                dp[i].extend(dp[sub_ans_index])
            if ans_size < len(dp[i]):
                ans_size = len(dp[i])
                ans_index = i
        return dp[ans_index]

File: 09_02_2024_Largest_Divisible_Subset.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

public class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> dp = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            dp.add(new ArrayList<>());
        }
        dp.get(n-1).add(nums[n-1]);
        int ans_size = 1, ans_index = n-1;
        for(int i = n-2; i >= 0; i--){
            int sub_ans_size = 0, sub_ans_index = -1;
            for(int j = i+1; j < n; j++){
                if(nums[j] % nums[i] == 0 && dp.get(j).size() > sub_ans_size){
                    sub_ans_size = dp.get(j).size();
                    sub_ans_index = j;
                }
            }
            dp.get(i).add(nums[i]);
            if(sub_ans_index != -1) dp.get(i).addAll(dp.get(sub_ans_index));
            if(ans_size < dp.get(i).size()){ 
                ans_size = dp.get(i).size();
                ans_index = i;
            }
        }
        return dp.get(ans_index);
    }
}

